Logging client

franchuterivera opened the pull request:
- Moving to log client exclusively and removed other support

- Adding checker to make sure expected messages are in log file

mfeurer commented on test/test_evaluation/test_train_evaluator.py:
```diff
@@ -63,6 +67,7 @@ def setUp(self):
         dummy_pred_files = [os.path.join(self.ev_path, str(n)) for n in range(100, 200)]
         dummy_cv_model_files = [os.path.join(self.ev_path, str(n)) for n in range(200, 300)]
         backend_mock = unittest.mock.Mock()
+        backend_mock.temporary_directory = tempfile.gettempdir()
```
Are all of these necessary given that the backend mock has a temporary directory itself?

franchuterivera replied:
In this particular case, the backend mock doesn't have a temporary directory and that is the reason why I added this. I have unified the backend mocks that could be unified under evaluation utils, but for sure it will be cleaner when we move to pytest fixtures.

mfeurer replied:
Thanks. Yes, I'm also looking forward to using pytest fixture, but I'm afraid that has to wait a bit more...

mfeurer commented on test/test_evaluation/test_test_evaluator.py:
```diff
@@ -43,6 +44,7 @@ def test_datasets(self):
 
             with self.subTest(testname):
                 backend_mock = unittest.mock.Mock(spec=Backend)
+                backend_mock.temporary_directory = tempfile.gettempdir()
```
Could the backend mock be generalized?

franchuterivera replied:
I have moved this to evaluation utils

mfeurer commented on test/test_evaluation/test_evaluation.py:
```diff
@@ -29,6 +30,9 @@ def safe_eval_success_mock(*args, **kwargs):
 
 
 class BackendMock(object):
+    def __init__(self):
```
Same as below, can this be generalized?

franchuterivera replied:
I have moved this to evaluation utils

mfeurer reviewed (commented):
First batch of review, but only this simple stuff so far...

mfeurer commented on autosklearn/evaluation/__init__.py:
```diff
@@ -101,7 +101,8 @@ def __init__(self, backend, autosklearn_seed, resampling_strategy, metric,
                  output_y_hat_optimization=True, include=None, exclude=None,
                  memory_limit=None, disable_file_output=False, init_params=None,
-                 budget_type=None, ta=False, pynisher_context='spawn', **resampling_strategy_args):
+                 budget_type=None, ta=False, port=9020,
```
The port is mandatory I guess?

franchuterivera replied:
Yes, I have now moved to mandatory ports

mfeurer commented on autosklearn/util/backend.py:
```diff
@@ -81,9 +82,19 @@ def __init__(self,
         self._output_directory = output_directory
-        self._logger = logging.get_logger(__name__)
+        # We create a dummy logger to start with
+        # Then, when a port is available we can create a
```
Could you please rephrase that this is passed from the outside?

franchuterivera replied:
done!

mfeurer replied:
This change is not pushed yet, right?

franchuterivera replied:
Sorry, I missed adding this filed. It has now been added.

mfeurer commented on test/test_automl/automl_utils.py:
```diff
@@ -55,3 +56,74 @@ def count_succeses(cv_results):
+
+        with open(self.logfile) as f:
+            content = f.readlines()
+        content = [x.strip() for x in content if re.search(r'[\w+]', x)]
```
Could you please do:
```
with open(self.logfile) as fh:
    content = [line.strip() for line in fh if re.search(r'[\w+]', line)]
```
which is shorter and easier to read. So this basically checks that a line has only unicode characters?

franchuterivera replied:
Done!

mfeurer commented on test/test_automl/automl_utils.py:
```diff
@@ -55,3 +56,74 @@ def count_succeses(cv_results):
+
+            # Pynisher call
+            match = re.search(
+                r'EnsembleBuilder]\s+Function called with argument:\s+\(\d+\.\d+, (\d+), \w+',
```
Don't you have to escape the `]`?

franchuterivera replied:
actually, if we escape it will fail because I am looking for ]. If you escape it, it assumes it is the bracket regexp that says any-of

mfeurer commented on test/test_automl/automl_utils.py:
```diff
@@ -55,3 +56,74 @@ def count_succeses(cv_results):
+
+            # Ensemble Builder actuall call
+            match = re.search(
+                r'EnsembleBuilder]\s+Starting iteration (\d+)',
```
here as well

franchuterivera replied:
DOne!

mfeurer reviewed (commented):
Part 2.

mfeurer commented on autosklearn/util/logging_.py:
```diff
@@ -66,7 +73,9 @@ def __getstate__(self) -> Dict[str, Any]:
         the self.logger field and only returns the logger name.
         """
-        return {'name': self.name}
+        return {'name': self.name, 'host': self.host, 'port': self.port,
```
Could you please put each key-value pair in a separate line?

franchuterivera replied:
Done!

mfeurer replied:
That one isn't pushed yet, right.

mfeurer reviewed (commented):
Review 3/3. 

Did you check if this works in singularity? Could you maybe add a test the log files are not created in the current working directory in case the logging is set up correctly?

franchuterivera commented on autosklearn/metalearning/metafeatures/metafeatures.py:
```diff
@@ -674,20 +674,26 @@ def _calculate(self, X, y, categorical):
             return accuracy / 5
         except scipy.linalg.LinAlgError as e:
-            self.logger.warning("LDA failed: %s Returned 0 instead!" % e)
+            if self.logger is not None:
```
Yes, that is the case. I will make it so that I change the testing code and always expect a logger.

franchuterivera commented on autosklearn/smbo.py:
```diff
@@ -222,6 +223,7 @@ def __init__(self, config_space, dataset_name,
                  smac_scenario_args=None,
                  get_smac_object_callback=None,
                  ensemble_callback: typing.Optional[EnsembleBuilderManager] = None,
+                 port=9020,
```
This port number is the default port logging value. I will make it so that it is required.

franchuterivera commented on autosklearn/evaluation/train_evaluator.py:
```diff
@@ -142,6 +142,7 @@ def _fit_with_budget(X_train, Y_train, budget, budget_type, logger, model, train
 
 class TrainEvaluator(AbstractEvaluator):
     def __init__(self, backend, queue, metric,
+                 port=9020,
```
This port number is the default port logging value. I will make it so that it is required.

franchuterivera commented on .pre-commit-config.yaml:
```diff
@@ -19,7 +19,7 @@ repos:
       - id: flake8
         name: flake8 auto-sklearn
-        files: autosklearn/*
+        files: autosklearn/.*
```
I got a warning in the new version saying that this is a regexp, not a glob. I did not perceive any change and with * is how openml has it. 

Don;t you also get the warning?

mfeurer replied:
Didn't get one yet, but it appears to still work given the flake failure in the current tests, so I'm fine with this.

franchuterivera commented on test/test_automl/automl_utils.py:
```diff
@@ -55,3 +56,74 @@ def count_succeses(cv_results):
+            if match:
+                iterations.append(int(match.group(1)))
+
+            # Ensemble Builder actuall call
```
Done!

franchuterivera commented on test/test_automl/automl_utils.py:
```diff
@@ -55,3 +56,74 @@ def count_succeses(cv_results):
+        # we have to count the start msg from pynisher
+        # and the return msg
+        # We expect the start msg to be something like:
+        # [DEBUG] [2020-11-26 19:22:42,160:EnsembleBuilder] Function called with argument: (61.
```
The rest of the expression varies depending on the result. That is the common part.

mfeurer replied:
Could you then please paste two or three examples here and add a comment that they differ? That'll make it easier in the future.

franchuterivera replied:
I added 2 examples per scenario for the future.

franchuterivera commented on test/test_automl/automl_utils.py:
```diff
@@ -55,3 +56,74 @@ def count_succeses(cv_results):
     )
+
+
+class AutoMLLogParser(object):
```
Done!

franchuterivera commented on test/test_automl/automl_utils.py:
```diff
@@ -55,3 +56,74 @@ def count_succeses(cv_results):
+    def count_tae_pynisher_calls(self):
+        # We expect the return msg to be something like:
+        # [DEBUG] [2020-11-26 19:22:39,558:pynisher] return value: (...
+        return_msgs = len([x for x in self.lines if re.search(
```
Done!

franchuterivera commented on test/test_automl/automl_utils.py:
```diff
@@ -55,3 +56,74 @@ def count_succeses(cv_results):
+
+        # We expect the return msg to be something like:
+        # [DEBUG] [2020-11-26 19:22:43,018:EnsembleBuilder] return value: (([{'Times...
+        return_msgs = len([x for x in self.lines if re.search(
```
Done!

mfeurer commented:
Thanks for the updates @franchuterivera the changes look good, but I'll wait with approving until all tests are green.

When looking at the tests I realized that there's a lot of output now and I'm not sure where it comes from, but it makes finding the failed tests quite hard. Do you know where it comes from and whether it can be reduced again?

franchuterivera commented:
Sorry, I also missed to add the sparse file and there was a bug here and there. 

I have fixed the output of the pytest.

I have added a new test to see what things are left behind by pytest to catch if we will be able to run singularity. This is likely going to fail if we miss some file. Once I see it I will fix this.

mfeurer commented on scripts/run_auto-sklearn_for_metadata_generation.py:
```diff
@@ -144,6 +145,7 @@
                                         include=include,
                                         metric=automl_arguments['metric'],
                                         cost_for_crash=get_cost_of_crash(automl_arguments['metric']),
+                                        port=logging.handlers.DEFAULT_TCP_LOGGING_PORT,
```
I'm just seeing this and there won't be any server running there so actually log this. Do you think we can only tear down the logging server in the `__del__` instead of also tearing it down at the end of `fit()`?

franchuterivera replied:
I think <__del__> is only called when you called del or when all references are removed right?

If for some reason any logger is still logging to the server, then references to the automl object will still be alive, so del would not be called. This can easily be triggered if you make the tae/exec in  smbo crash at start. Then the logger will never return if this logger clean up (with time out) doesn't trigger.

Do you mean this for the metalearning testing? Maybe it makes sense to make it so that if port is none we default to logging.getLogger with a warning message?

mfeurer replied:
> I think <del> is only called when you called del or when all references are removed right?

Yes, I think so, too.

> If for some reason any logger is still logging to the server, then references to the automl object will still be alive, so del would not be called. This can easily be triggered if you make the tae/exec in smbo crash at start. Then the logger will never return if this logger clean up (with time out) doesn't trigger.

I see.

> Do you mean this for the metalearning testing? Maybe it makes sense to make it so that if port is none we default to logging.getLogger with a warning message?

Both for testing and practical use. I think the idea of falling back to logging.getLogger is a great idea.

mfeurer reviewed (commented):
This now looks good to me. However, there are still several files under `autosklearn.metalearning` which have `logger=None` in the constructor. Could you please make this a mandatory argument?

mfeurer commented on test/test_automl/test_automl.py:
```diff
@@ -334,6 +329,52 @@ def test_automl_outputs(backend, dask_client):
+        # Make sure that running jobs are properly tracked. Killed runs do not always
+        # print the return value to the log file (yet the run history has this information)
+        if run_value.status == StatusType.SUCCESS:
+            # Success is not sufficient to guarantee a return message in the log file
```
Just checking, if the run doesn't print `r'pynisher]\s+return value:\s+'` it should print either or ` "Your function call closed the pipe prematurely -> Subprocess probably got an uncatchable signal."` or `"Something else went wrong, sorry."` in the controlling process, so we could actually check that, right?

franchuterivera replied:
If a failure happens [here](https://github.com/automl/pynisher/blob/beb76e4394d364efb4847809a2f06ec5e1f18d54/pynisher/limit_function_call.py#L154) we will get a double count message, because while sending the pipe message it could break. 

However, we can make this deterministic and add a message at the end of the pynisher function call (inside pynisher)? For instance [here](https://github.com/automl/pynisher/blob/beb76e4394d364efb4847809a2f06ec5e1f18d54/pynisher/limit_function_call.py#L280)

mfeurer replied:
But aren't the two error messages I mentioned exactly the error messages showing up in case of a failure to write to the pipe?

mfeurer commented on autosklearn/evaluation/__init__.py:
```diff
@@ -261,8 +266,15 @@ def run(
         if self.init_params is not None:
             init_params.update(self.init_params)
 
+        if self.port is None:
```
The handling of `port is None` is different for this logger and the logger above. Can this be unified?

franchuterivera replied:
The only reason this is not unified is to get a different name for the TAE and pynisher call. Is it ok if we only get TAE in the message name?

mfeurer replied:
Sorry, I meant that in one case it check if the port is None, in the other it doesn't check.

franchuterivera commented:
I have created #1039 to make testing more restrictive after the running task left out at the end of smac is resolved
